slapo/verify.py

```diff
@@ -0,0 +1,106 @@
+        self.device = device
+
+    def __enter__(self):
+        self.original_trace = sys.gettrace()
+
+        # pylint: disable=unused-argument
+        def trace_calls(frame, event, arg):
+            if event == "call":
+                code = frame.f_code
+                function_name = code.co_name
+                # local_sch = frame.f_locals.get("sch")
+
+                if function_name == "apply":
+                    # This part is useful only when we need to get the model from the schedule
+                    # (the schedule is not passed in as an argument)
+                    for _, value in frame.f_globals.items():
+                        cls_name = getattr(value, "__name__", None)
+                        if cls_name in PRIMITIVES_NAMES and value.is_verifiable():
+                            logger.info("Verifying %s...", cls_name, ranks=0)
+                            break
+
+            return trace_calls
+
+        sys.settrace(trace_calls)
+        return self
+
+    def __exit__(self, *exc):
+        """Verify the correctness of the schedule.
+        TODO: Support backward verification
+        """
+        # 1. Build the original model with random weights
+        named_params = self.original_sch.mod.named_parameters()
+        is_initialized = named_params.__next__()[1].device != torch.device("meta")
+        original_mod, _ = build(self.original_sch, init_weights=not is_initialized)
+        original_mod = original_mod.to(self.device)
+        #    Broadcast the original model from rank 0 to other ranks
+        original_state_dict = original_mod.state_dict()
+        if dist.is_initialized():
+            for param_name in original_state_dict:
+                dist.broadcast(
+                    original_state_dict[param_name], src=0, group=self.sch.group
+                )
+        # 2. Get the transformed model from the schedule
+        #    Copy it and build a new schedule to prevent the original schedule from being modified
+        copied_mod = copy.deepcopy(self.sch.mod)
+        # copy original attributes
+        for param_name, param in self.sch.mod.named_parameters():
+            if hasattr(param, "orig_shape"):
+                copied_mod.get_parameter(param_name).orig_shape = param.orig_shape
+        new_sch = create_schedule(copied_mod)
+        # 3. Use original weights to initialize the new model
+        #    Notice init_weights is called before actual sharding, so we only need to
+        #    assign the original weights to the corresponding modules
+
+        def init_weights(mod, path):
+            if hasattr(mod, "weight") and mod.weight is not None:
+                mod.weight = nn.Parameter(
+                    original_state_dict[f"{path}.weight"].detach()
+                )
+            if hasattr(mod, "bias") and mod.bias is not None:
+                mod.bias = nn.Parameter(original_state_dict[f"{path}.bias"].detach())
```

Comment:
Since this is just syncing parameters, can we make it more systematic, such as the following?

```python
for name, _ in mod.named_parameters():
    setattr(mod, name, nn.Parameter(original_state_dict[f"{path}.{name}"].detach()))
```